sycl/doc/extensions/proposed/sycl_ext_oneapi_barrier.asciidoc

```diff
@@ -0,0 +1,458 @@
+_Preconditions_: If `Scope` is `memory_scope::work_group`, the calling thread
+of execution must be a work-item belonging to the work-group that will use the
+barrier. If `Scope` is `memory_scope::device`, the calling thread of execution
+must be executing on the device where the barrier will be used. If `Scope` is
+`memory_scope::system`, the calling thread of execution can be any host or
+device thread.
+
+_Effects_: Initializes the barrier with the given expected count and completion
+function, and begins the first phase of the barrier.
+
+[source,c++]
+----
+[[nodiscard]] arrival_token arrive(ptrdiff_t update = 1);
+----
+_Effects_: The calling thread of execution arrives at the barrier and decreases
+the expected count by `update`.
+
+_Returns_: An `arrival_token` that can be passed to `wait`.
+
+[source,c++]
+----
+void wait(arrival_token&& arrival) const;
+----
+_Effects_: Blocks the calling thread of execution until the end of the barrier
+phase associated with `arrival`.
+
+[NOTE]
+====
+Because calling `wait` blocks the calling thread of execution, it may prevent
+forward progress. Unlike `group_barrier`, the member functions of `barrier` do
+not provide additional scheduling guarantees; it is a user's responsibility to
+ensure that calls to `wait` are compatible with the forward progress guarantees
+provided by an implementation.
+====
+
+[source,c++]
+----
+void arrive_and_wait();
+----
+_Effects_: Equivalent to `wait(arrive())`.
+
+[source,c++]
+----
+void arrive_and_drop();
+----
+_Effects_: The calling thread of execution arrives at the barrier, decreases
+the number of threads of execution expected in the next phase, and decreases
+the expected count of the current phase by 1.
+
+
+==== Usage examples
+
+As noted above, `barrier` objects must be allocated in memory that is visible
+to all of the work-items using the barrier. There are many ways to satisfy this
+condition. The examples in this section demonstrate some common ways to
+allocate and construct barrier objects for different scopes.
+
+
+===== Work-group scope
+
+Barriers at work-group scope can be allocated in group-local memory. The
+simplest way to allocate and construct a `barrier` object in group-local memory
+is to use the
+link:../supported/sycl_ext_oneapi_local_memory.asciidoc[sycl_ext_oneapi_local_memory]
+extension.
+
+[source,c++]
+----
+using work_group_barrier = syclex::barrier<sycl::memory_scope::work_group>;
+
+q.parallel_for(..., [=](sycl::nd_item it) {
+
+  // Allocate memory for and construct the barrier
+  auto* bar = sycl::ext::oneapi::group_local_memory<work_group_barrier>(it.get_group(), nthreads);
+
+  // Use the barrier
+  bar->arrive_and_wait();
+
+  // The barrier is automatically destructed and the memory is freed upon kernel completion
+
+}).wait();
+----
+
+
+===== Device scope
+
+Barriers at device scope must be allocated in global memory (including various
+kinds of USM). It is recommended that device scope barriers are initialized on
```

Comment:
I think it is not strictly necessary for the barrier to be allocated in global memory.  For example, I assume it is possible to allocate the barrier in SLM if the only threads synchronizing on the barrier are in the same work-group.  (It would be overkill to define the barrier as `memory_scope::device` in this case, but I assume it still legal.)

Could we weaken this language to say "Barriers at device scope are normally allocated in global memory ..."?  Same comment below for the "system scope" example.

Reply:
I think you're right, but I have a nagging doubt that some implementations may complain about this.  If a device barrier is guaranteed to be allocated in global memory, then any atomics used by the barrier implementation can assume the global address space.  If a device barrier can be allocated anywhere, the implementation needs to use the generic address space or otherwise somehow specialize the barrier behavior.

I've made the change in cc7eaf559699a759c9cde1586e3113f9c1479bda, because my implementation fear may prove to be unfounded.  This is similar to how we handled the concern raised in https://github.com/intel/llvm/pull/9186#discussion_r1186148501.  We may need to revisit these sorts of restrictions when we have implementation experience.